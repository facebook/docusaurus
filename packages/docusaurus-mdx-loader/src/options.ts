/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {MDXOptions, SimpleProcessors} from './processor';
import type {MarkdownConfig} from '@docusaurus/types';
import type {ResolveMarkdownLink} from './remark/resolveMarkdownLinks';
import type {PromiseWithResolvers} from './utils';

export type Options = Partial<MDXOptions> & {
  markdownConfig: MarkdownConfig;
  staticDirs: string[];
  siteDir: string;
  isMDXPartial?: (filePath: string) => boolean;
  isMDXPartialFrontMatterWarningDisabled?: boolean;
  removeContentTitle?: boolean;
  metadataPath?: (filePath: string) => string;
  createAssets?: (metadata: {
    filePath: string;
    frontMatter: {[key: string]: unknown};
  }) => {[key: string]: unknown};
  resolveMarkdownLink?: ResolveMarkdownLink;

  // Will usually be created by "createMDXLoaderItem"
  processors?: SimpleProcessors;
  crossCompilerCache?: Map<string, CrossCompilerCacheEntry>; // MDX => Promise<JSX> cache
};

type CrossCompilerCacheEntry = PromiseWithResolvers<string>;
