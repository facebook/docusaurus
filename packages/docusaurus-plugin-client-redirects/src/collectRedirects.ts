/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import _ from 'lodash';
import logger from '@docusaurus/logger';
import {
  applyTrailingSlash,
  addTrailingSlash,
  removeTrailingSlash,
} from '@docusaurus/utils-common';
import {
  createFromExtensionsRedirects,
  createToExtensionsRedirects,
} from './extensionRedirects';
import {validateRedirect} from './redirectValidation';
import type {PluginOptions, RedirectOption} from './options';
import type {PluginContext, RedirectItem} from './types';

export default function collectRedirects(
  pluginContext: PluginContext,
  trailingSlash: boolean | undefined,
): RedirectItem[] {
  // For each plugin config option, create the appropriate redirects
  const redirects = [
    ...createFromExtensionsRedirects(
      pluginContext.relativeRoutesPaths,
      pluginContext.options.fromExtensions,
    ),
    ...createToExtensionsRedirects(
      pluginContext.relativeRoutesPaths,
      pluginContext.options.toExtensions,
    ),
    ...createRedirectsOptionRedirects(pluginContext.options.redirects),
    ...createCreateRedirectsOptionRedirects(
      pluginContext.relativeRoutesPaths,
      pluginContext.options.createRedirects,
    ),
  ].map((redirect) => ({
    ...redirect,
    // Given a redirect with `to: "/abc"` and `trailingSlash` enabled:
    //
    // - We don't want to reject `to: "/abc"`, as that unambiguously points to
    // `/abc/` now;
    // - We want to redirect `to: /abc/` without the user having to change all
    // her redirect plugin options
    //
    // It should be easy to toggle `trailingSlash` option without having to
    // change other configs
    to: applyTrailingSlash(redirect.to, {
      trailingSlash,
      baseUrl: pluginContext.baseUrl,
    }),
  }));

  validateCollectedRedirects(redirects, pluginContext);
  return filterUnwantedRedirects(redirects, pluginContext);
}

function validateCollectedRedirects(
  redirects: RedirectItem[],
  pluginContext: PluginContext,
) {
  const redirectValidationErrors = redirects
    .map((redirect) => {
      try {
        validateRedirect(redirect);
        return undefined;
      } catch (err) {
        return (err as Error).message;
      }
    })
    .filter(Boolean);
  if (redirectValidationErrors.length > 0) {
    throw new Error(
      `Some created redirects are invalid:
- ${redirectValidationErrors.join('\n- ')}
`,
    );
  }

  const allowedToPaths = pluginContext.relativeRoutesPaths.map((p) =>
    decodeURI(p),
  );
  const toPaths = redirects
    .map((redirect) => redirect.to)
    // We now allow "to" to contain any string
    // We only do this "broken redirect" check from to that looks like pathnames
    // note: we allow querystring/anchors
    // See https://github.com/facebook/docusaurus/issues/6845
    .map((to) => {
      if (to.startsWith('/')) {
        try {
          return decodeURI(new URL(to, 'https://example.com').pathname);
        } catch (e) {}
      }
      return undefined;
    })
    .filter((to): to is string => typeof to !== 'undefined');

  const trailingSlashConfig = pluginContext.siteConfig.trailingSlash;
  // Key is the path, value is whether a valid toPath with a different trailing
  // slash exists; if the key doesn't exist it means it's valid
  const differByTrailSlash = new Map(toPaths.map((path) => [path, false]));
  allowedToPaths.forEach((toPath) => {
    if (differByTrailSlash.has(toPath)) {
      differByTrailSlash.delete(toPath);
    } else if (differByTrailSlash.has(removeTrailingSlash(toPath))) {
      if (trailingSlashConfig === true) {
        differByTrailSlash.set(removeTrailingSlash(toPath), true);
      } else {
        differByTrailSlash.delete(removeTrailingSlash(toPath));
      }
    } else if (differByTrailSlash.has(addTrailingSlash(toPath))) {
      if (trailingSlashConfig === false) {
        differByTrailSlash.set(addTrailingSlash(toPath), true);
      } else {
        differByTrailSlash.delete(addTrailingSlash(toPath));
      }
    }
  });
  if (differByTrailSlash.size > 0) {
    const errors = Array.from(differByTrailSlash.entries());

    let message =
      'You are trying to create client-side redirections to invalid paths.\n';

    const [trailingSlashIssues, invalidPaths] = _.partition(
      errors,
      ([, differ]) => differ,
    );

    if (trailingSlashIssues.length) {
      message += `
These paths do exist, but because you have explicitly set trailingSlash=${trailingSlashConfig}, you need to write the path ${
        trailingSlashConfig ? 'with trailing slash' : 'without trailing slash'
      }:
- ${trailingSlashIssues.map(([p]) => p).join('\n- ')}
`;
    }

    if (invalidPaths.length) {
      message += `
These paths are redirected to but do not exist:
- ${invalidPaths.map(([p]) => p).join('\n- ')}

Valid paths you can redirect to:
- ${allowedToPaths.join('\n- ')}
`;
    }

    throw new Error(message);
  }
}

function filterUnwantedRedirects(
  redirects: RedirectItem[],
  pluginContext: PluginContext,
): RedirectItem[] {
  // We don't want to create the same redirect twice, since that would lead to
  // writing the same html redirection file twice.
  Object.entries(_.groupBy(redirects, (redirect) => redirect.from)).forEach(
    ([from, groupedFromRedirects]) => {
      if (groupedFromRedirects.length > 1) {
        logger.report(
          pluginContext.siteConfig.onDuplicateRoutes,
        )`name=${'@docusaurus/plugin-client-redirects'}: multiple redirects are created with the same "from" pathname: path=${from}
It is not possible to redirect the same pathname to multiple destinations:${groupedFromRedirects.map(
          (r) => JSON.stringify(r),
        )}`;
      }
    },
  );
  const collectedRedirects = _.uniqBy(redirects, (redirect) => redirect.from);

  const {false: newRedirects = [], true: redirectsOverridingExistingPath = []} =
    _.groupBy(collectedRedirects, (redirect) =>
      pluginContext.relativeRoutesPaths.includes(redirect.from),
    );
  if (redirectsOverridingExistingPath.length > 0) {
    logger.report(
      pluginContext.siteConfig.onDuplicateRoutes,
    )`name=${'@docusaurus/plugin-client-redirects'}: some redirects would override existing paths, and will be ignored:${redirectsOverridingExistingPath.map(
      (r) => JSON.stringify(r),
    )}`;
  }
  return newRedirects;
}

function createRedirectsOptionRedirects(
  redirectsOption: PluginOptions['redirects'],
): RedirectItem[] {
  // For convenience, user can use a string or a string[]
  function optionToRedirects(option: RedirectOption): RedirectItem[] {
    if (typeof option.from === 'string') {
      return [{from: option.from, to: option.to}];
    }
    return option.from.map((from) => ({from, to: option.to}));
  }

  return redirectsOption.flatMap(optionToRedirects);
}

// Create redirects from the "createRedirects" fn provided by the user
function createCreateRedirectsOptionRedirects(
  paths: string[],
  createRedirects: PluginOptions['createRedirects'],
): RedirectItem[] {
  function createPathRedirects(path: string): RedirectItem[] {
    const fromsMixed: string | string[] = createRedirects?.(path) ?? [];

    const froms: string[] =
      typeof fromsMixed === 'string' ? [fromsMixed] : fromsMixed;

    return froms.map((from) => ({from, to: path}));
  }

  return paths.flatMap(createPathRedirects);
}
