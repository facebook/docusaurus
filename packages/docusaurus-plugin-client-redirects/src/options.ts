/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {Joi, PathnameSchema} from '@docusaurus/utils-validation';
import type {OptionValidationContext} from '@docusaurus/types';

export type RedirectOption = {
  /** Pathname of an existing Docusaurus page */
  to: string;
  /** Pathname of the new page(s) we should create */
  from: string | string[];
};

export type PluginOptions = {
  /** Plugin ID. */
  id: string;
  /** The extensions to be removed from the route after redirecting. */
  fromExtensions: string[];
  /** The extensions to be appended to the route after redirecting. */
  toExtensions: string[];
  /** The list of redirect rules, each one with multiple `from`s â†’ one `to`. */
  redirects: RedirectOption[];
  /**
   * A callback to create a redirect rule. Docusaurus query this callback
   * against every path it has created, and use its return value to output more
   * paths.
   * @returns All the paths from which we should redirect to `path`
   */
  createRedirects?: (
    /** An existing Docusaurus route path */
    path: string,
  ) => string[] | string | null | undefined;
};

export type Options = Partial<PluginOptions>;

export const DEFAULT_OPTIONS: Partial<PluginOptions> = {
  fromExtensions: [],
  toExtensions: [],
  redirects: [],
};

const RedirectPluginOptionValidation = Joi.object<RedirectOption>({
  from: Joi.alternatives().try(
    PathnameSchema.required(),
    Joi.array().items(PathnameSchema.required()),
  ),
  to: Joi.string().required(),
});

const isString = Joi.string().required().not(null);

const UserOptionsSchema = Joi.object<PluginOptions>({
  fromExtensions: Joi.array()
    .items(isString)
    .default(DEFAULT_OPTIONS.fromExtensions),
  toExtensions: Joi.array()
    .items(isString)
    .default(DEFAULT_OPTIONS.toExtensions),
  redirects: Joi.array()
    .items(RedirectPluginOptionValidation)
    .default(DEFAULT_OPTIONS.redirects),
  createRedirects: Joi.function().maxArity(1),
}).default(DEFAULT_OPTIONS);

export function validateOptions({
  validate,
  options: userOptions,
}: OptionValidationContext<Options | undefined, PluginOptions>): PluginOptions {
  return validate(UserOptionsSchema, userOptions);
}
