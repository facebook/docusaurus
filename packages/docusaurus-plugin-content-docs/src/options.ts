/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import logger from '@docusaurus/logger';
import {
  Joi,
  RemarkPluginsSchema,
  RehypePluginsSchema,
  RecmaPluginsSchema,
  AdmonitionsSchema,
  RouteBasePathSchema,
  URISchema,
} from '@docusaurus/utils-validation';
import {GlobExcludeDefault} from '@docusaurus/utils';
import {DefaultSidebarItemsGenerator} from './sidebars/generator';
import {
  DefaultNumberPrefixParser,
  DisabledNumberPrefixParser,
} from './numberPrefix';
import type {OptionValidationContext} from '@docusaurus/types';
import type {PluginOptions, Options} from '@docusaurus/plugin-content-docs';

export const DEFAULT_OPTIONS: Omit<PluginOptions, 'id' | 'sidebarPath'> = {
  path: 'docs', // Path to data on filesystem, relative to site dir.
  routeBasePath: 'docs', // URL Route.
  tagsBasePath: 'tags', // URL Tags Route.
  include: ['**/*.{md,mdx}'], // Extensions to include.
  exclude: GlobExcludeDefault,
  sidebarItemsGenerator: DefaultSidebarItemsGenerator,
  numberPrefixParser: DefaultNumberPrefixParser,
  docsRootComponent: '@theme/DocsRoot',
  docVersionRootComponent: '@theme/DocVersionRoot',
  docRootComponent: '@theme/DocRoot',
  docItemComponent: '@theme/DocItem',
  docTagDocListComponent: '@theme/DocTagDocListPage',
  docTagsListComponent: '@theme/DocTagsListPage',
  docCategoryGeneratedIndexComponent: '@theme/DocCategoryGeneratedIndexPage',
  remarkPlugins: [],
  rehypePlugins: [],
  recmaPlugins: [],
  beforeDefaultRemarkPlugins: [],
  beforeDefaultRehypePlugins: [],
  showLastUpdateTime: false,
  showLastUpdateAuthor: false,
  admonitions: true,
  includeCurrentVersion: true,
  disableVersioning: false,
  lastVersion: undefined,
  versions: {},
  editCurrentVersion: false,
  editLocalizedFiles: false,
  sidebarCollapsible: true,
  sidebarCollapsed: true,
  breadcrumbs: true,
  onInlineTags: 'warn',
  tags: undefined,
};

const VersionOptionsSchema = Joi.object({
  path: Joi.string().allow('').optional(),
  label: Joi.string().optional(),
  banner: Joi.string().equal('none', 'unreleased', 'unmaintained').optional(),
  badge: Joi.boolean().optional(),
  className: Joi.string().optional(),
  noIndex: Joi.boolean().optional(),
});

const VersionsOptionsSchema = Joi.object()
  .pattern(Joi.string().required(), VersionOptionsSchema)
  .default(DEFAULT_OPTIONS.versions);

const OptionsSchema = Joi.object<PluginOptions>({
  path: Joi.string().default(DEFAULT_OPTIONS.path),
  editUrl: Joi.alternatives().try(URISchema, Joi.function()),
  editCurrentVersion: Joi.boolean().default(DEFAULT_OPTIONS.editCurrentVersion),
  editLocalizedFiles: Joi.boolean().default(DEFAULT_OPTIONS.editLocalizedFiles),
  routeBasePath: RouteBasePathSchema.default(DEFAULT_OPTIONS.routeBasePath),
  tagsBasePath: Joi.string().default(DEFAULT_OPTIONS.tagsBasePath),
  // @ts-expect-error: deprecated
  homePageId: Joi.any().forbidden().messages({
    'any.unknown':
      'The docs plugin option homePageId is not supported anymore. To make a doc the "home", please add "slug: /" in its front matter. See: https://docusaurus.io/docs/next/docs-introduction#home-page-docs',
  }),

  include: Joi.array().items(Joi.string()).default(DEFAULT_OPTIONS.include),
  exclude: Joi.array().items(Joi.string()).default(DEFAULT_OPTIONS.exclude),
  sidebarPath: Joi.alternatives().try(
    Joi.boolean().invalid(true),
    Joi.string(),
  ),
  sidebarItemsGenerator: Joi.function().default(
    () => DEFAULT_OPTIONS.sidebarItemsGenerator,
  ),
  sidebarCollapsible: Joi.boolean().default(DEFAULT_OPTIONS.sidebarCollapsible),
  sidebarCollapsed: Joi.boolean().default(DEFAULT_OPTIONS.sidebarCollapsed),
  numberPrefixParser: Joi.alternatives()
    .try(
      Joi.function(),
      // Convert boolean values to functions
      Joi.alternatives().conditional(Joi.boolean(), {
        then: Joi.custom((val: boolean) =>
          val ? DefaultNumberPrefixParser : DisabledNumberPrefixParser,
        ),
      }),
    )
    .default(() => DEFAULT_OPTIONS.numberPrefixParser),
  docsRootComponent: Joi.string().default(DEFAULT_OPTIONS.docsRootComponent),
  docVersionRootComponent: Joi.string().default(
    DEFAULT_OPTIONS.docVersionRootComponent,
  ),
  docRootComponent: Joi.string().default(DEFAULT_OPTIONS.docRootComponent),
  docItemComponent: Joi.string().default(DEFAULT_OPTIONS.docItemComponent),
  docTagsListComponent: Joi.string().default(
    DEFAULT_OPTIONS.docTagsListComponent,
  ),
  docTagDocListComponent: Joi.string().default(
    DEFAULT_OPTIONS.docTagDocListComponent,
  ),
  docCategoryGeneratedIndexComponent: Joi.string().default(
    DEFAULT_OPTIONS.docCategoryGeneratedIndexComponent,
  ),
  remarkPlugins: RemarkPluginsSchema.default(DEFAULT_OPTIONS.remarkPlugins),
  rehypePlugins: RehypePluginsSchema.default(DEFAULT_OPTIONS.rehypePlugins),
  recmaPlugins: RecmaPluginsSchema.default(DEFAULT_OPTIONS.recmaPlugins),
  beforeDefaultRemarkPlugins: RemarkPluginsSchema.default(
    DEFAULT_OPTIONS.beforeDefaultRemarkPlugins,
  ),
  beforeDefaultRehypePlugins: RehypePluginsSchema.default(
    DEFAULT_OPTIONS.beforeDefaultRehypePlugins,
  ),
  admonitions: AdmonitionsSchema.default(DEFAULT_OPTIONS.admonitions),
  showLastUpdateTime: Joi.bool().default(DEFAULT_OPTIONS.showLastUpdateTime),
  showLastUpdateAuthor: Joi.bool().default(
    DEFAULT_OPTIONS.showLastUpdateAuthor,
  ),
  includeCurrentVersion: Joi.bool().default(
    DEFAULT_OPTIONS.includeCurrentVersion,
  ),
  onlyIncludeVersions: Joi.array().items(Joi.string().required()).optional(),
  disableVersioning: Joi.bool().default(DEFAULT_OPTIONS.disableVersioning),
  lastVersion: Joi.string().optional(),
  versions: VersionsOptionsSchema,
  breadcrumbs: Joi.bool().default(DEFAULT_OPTIONS.breadcrumbs),
  onInlineTags: Joi.string()
    .equal('ignore', 'log', 'warn', 'throw')
    .default(DEFAULT_OPTIONS.onInlineTags),
  tags: Joi.string()
    .disallow('')
    .allow(null, false)
    .default(() => DEFAULT_OPTIONS.tags),
});

export function validateOptions({
  validate,
  options: userOptions,
}: OptionValidationContext<Options, PluginOptions>): PluginOptions {
  let options = userOptions;

  if (options.sidebarCollapsible === false) {
    // When sidebarCollapsible=false and sidebarCollapsed=undefined, we don't
    // want to have the inconsistency warning. We let options.sidebarCollapsible
    // become the default value for options.sidebarCollapsed
    if (typeof options.sidebarCollapsed === 'undefined') {
      options = {
        ...options,
        sidebarCollapsed: false,
      };
    }
    if (options.sidebarCollapsed) {
      logger.warn`The docs plugin config is inconsistent. It does not make sense to use code=${'sidebarCollapsible: false'} and code=${'sidebarCollapsed: true'} at the same time. code=${'sidebarCollapsed: true'} will be ignored.`;
      options = {
        ...options,
        sidebarCollapsed: false,
      };
    }
  }

  const normalizedOptions = validate(OptionsSchema, options);

  return normalizedOptions;
}
