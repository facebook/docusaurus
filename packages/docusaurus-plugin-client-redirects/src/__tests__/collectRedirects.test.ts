/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {removeTrailingSlash} from '@docusaurus/utils-common';
import {normalizePluginOptions} from '@docusaurus/utils-validation';
import collectRedirects from '../collectRedirects';
import {validateOptions} from '../options';
import type {DocusaurusConfig} from '@docusaurus/types';
import type {Options} from '../options';
import type {PluginContext} from '../types';

function createTestPluginContext(
  options?: Options,
  relativeRoutesPaths: string[] = [],
  siteConfig: Partial<DocusaurusConfig> = {},
): PluginContext {
  return {
    outDir: '/tmp',
    baseUrl: 'https://docusaurus.io',
    relativeRoutesPaths,
    options: validateOptions({validate: normalizePluginOptions, options}),
    siteConfig: {onDuplicateRoutes: 'warn', ...siteConfig} as DocusaurusConfig,
  };
}

describe('collectRedirects', () => {
  it('collects no redirect for undefined config', () => {
    expect(
      collectRedirects(
        createTestPluginContext(undefined, ['/', '/path']),
        undefined,
      ),
    ).toEqual([]);
  });

  it('collects no redirect for empty config', () => {
    expect(collectRedirects(createTestPluginContext({}), undefined)).toEqual(
      [],
    );
  });

  it('collects redirects from html/exe extension', () => {
    expect(
      collectRedirects(
        createTestPluginContext(
          {
            fromExtensions: ['html', 'exe'],
          },
          ['/', '/somePath', '/otherPath.html'],
        ),
        undefined,
      ),
    ).toEqual([
      {
        from: '/somePath.html',
        to: '/somePath',
      },
      {
        from: '/somePath.exe',
        to: '/somePath',
      },
    ]);
  });

  it('collects redirects to html/exe extension', () => {
    expect(
      collectRedirects(
        createTestPluginContext(
          {
            toExtensions: ['html', 'exe'],
          },
          ['/', '/somePath', '/otherPath.html'],
        ),
        undefined,
      ),
    ).toEqual([
      {
        from: '/otherPath',
        to: '/otherPath.html',
      },
    ]);
  });

  it('collects redirects from plugin option redirects', () => {
    expect(
      collectRedirects(
        createTestPluginContext(
          {
            redirects: [
              {
                from: '/someLegacyPath',
                to: '/somePath',
              },
              {
                from: '/someLegacyPath2',
                to: '/some Path2',
              },
              {
                from: '/someLegacyPath3',
                to: '/some%20Path3',
              },
              {
                from: ['/someLegacyPathArray1', '/someLegacyPathArray2'],
                to: '/',
              },

              {
                from: '/localQS',
                to: '/somePath?a=1&b=2',
              },
              {
                from: '/localAnchor',
                to: '/somePath#anchor',
              },
              {
                from: '/localQSAnchor',
                to: '/somePath?a=1&b=2#anchor',
              },

              {
                from: '/absolute',
                to: 'https://docusaurus.io/somePath',
              },
              {
                from: '/absoluteQS',
                to: 'https://docusaurus.io/somePath?a=1&b=2',
              },
              {
                from: '/absoluteAnchor',
                to: 'https://docusaurus.io/somePath#anchor',
              },
              {
                from: '/absoluteQSAnchor',
                to: 'https://docusaurus.io/somePath?a=1&b=2#anchor',
              },
            ],
          },
          ['/', '/somePath', '/some%20Path2', '/some Path3'],
        ),
        undefined,
      ),
    ).toEqual([
      {
        from: '/someLegacyPath',
        to: '/somePath',
      },
      {
        from: '/someLegacyPath2',
        to: '/some Path2',
      },
      {
        from: '/someLegacyPath3',
        to: '/some%20Path3',
      },
      {
        from: '/someLegacyPathArray1',
        to: '/',
      },
      {
        from: '/someLegacyPathArray2',
        to: '/',
      },
      {
        from: '/localQS',
        to: '/somePath?a=1&b=2',
      },
      {
        from: '/localAnchor',
        to: '/somePath#anchor',
      },
      {
        from: '/localQSAnchor',
        to: '/somePath?a=1&b=2#anchor',
      },

      {
        from: '/absolute',
        to: 'https://docusaurus.io/somePath',
      },
      {
        from: '/absoluteQS',
        to: 'https://docusaurus.io/somePath?a=1&b=2',
      },
      {
        from: '/absoluteAnchor',
        to: 'https://docusaurus.io/somePath#anchor',
      },
      {
        from: '/absoluteQSAnchor',
        to: 'https://docusaurus.io/somePath?a=1&b=2#anchor',
      },
    ]);
  });

  it('collects redirects from plugin option redirects with trailingSlash=true', () => {
    expect(
      collectRedirects(
        createTestPluginContext(
          {
            redirects: [
              {
                from: '/someLegacyPath',
                to: '/somePath',
              },
              {
                from: ['/someLegacyPathArray1', '/someLegacyPathArray2'],
                to: '/',
              },
            ],
          },
          ['/', '/somePath/'],
        ),
        true,
      ),
    ).toEqual([
      {
        from: '/someLegacyPath',
        to: '/somePath/',
      },
      {
        from: '/someLegacyPathArray1',
        to: '/',
      },
      {
        from: '/someLegacyPathArray2',
        to: '/',
      },
    ]);
  });

  it('collects redirects from plugin option redirects with trailingSlash=false', () => {
    expect(
      collectRedirects(
        createTestPluginContext(
          {
            redirects: [
              {
                from: '/someLegacyPath',
                to: '/somePath/',
              },
              {
                from: ['/someLegacyPathArray1', '/someLegacyPathArray2'],
                to: '/',
              },
            ],
          },
          ['/', '/somePath'],
        ),
        false,
      ),
    ).toEqual([
      {
        from: '/someLegacyPath',
        to: '/somePath',
      },
      {
        from: '/someLegacyPathArray1',
        to: '/',
      },
      {
        from: '/someLegacyPathArray2',
        to: '/',
      },
    ]);
  });

  it('throw if plugin option redirects contain invalid to paths', () => {
    expect(() =>
      collectRedirects(
        createTestPluginContext(
          {
            redirects: [
              {
                from: '/someLegacyPath',
                to: '/',
              },
              {
                from: '/someLegacyPath',
                to: '/this/path/does/not/exist2',
              },
              {
                from: '/someLegacyPath',
                to: '/this/path/does/not/exist3',
              },
              {
                from: '/someLegacyPath',
                to: '/this/path/does/not/exist4?a=b#anchor',
              },
            ],
          },
          ['/', '/someExistingPath', '/anotherExistingPath'],
        ),
        undefined,
      ),
    ).toThrowErrorMatchingSnapshot();
  });

  it('tolerates mismatched trailing slash if option is undefined', () => {
    expect(
      collectRedirects(
        createTestPluginContext(
          {
            redirects: [
              {
                from: '/someLegacyPath',
                to: '/somePath',
              },
            ],
          },
          ['/', '/somePath/'],
          {trailingSlash: undefined},
        ),
        undefined,
      ),
    ).toEqual([
      {
        from: '/someLegacyPath',
        to: '/somePath',
      },
    ]);
  });

  it('throw if plugin option redirects contain to paths with mismatching trailing slash', () => {
    expect(() =>
      collectRedirects(
        createTestPluginContext(
          {
            redirects: [
              {
                from: '/someLegacyPath',
                to: '/someExistingPath/',
              },
            ],
          },
          ['/', '/someExistingPath', '/anotherExistingPath'],
          {trailingSlash: false},
        ),
        undefined,
      ),
    ).toThrowErrorMatchingSnapshot();

    expect(() =>
      collectRedirects(
        createTestPluginContext(
          {
            redirects: [
              {
                from: '/someLegacyPath',
                to: '/someExistingPath',
              },
            ],
          },
          ['/', '/someExistingPath/', '/anotherExistingPath/'],
          {trailingSlash: true},
        ),
        undefined,
      ),
    ).toThrowErrorMatchingSnapshot();
  });

  it('collects redirects with custom redirect creator', () => {
    expect(
      collectRedirects(
        createTestPluginContext(
          {
            createRedirects: (routePath) => [
              `${removeTrailingSlash(routePath)}/some/path/suffix1`,
              `${removeTrailingSlash(routePath)}/some/other/path/suffix2`,
            ],
          },
          ['/', '/testPath', '/otherPath.html'],
        ),
        undefined,
      ),
    ).toEqual([
      {
        from: '/some/path/suffix1',
        to: '/',
      },
      {
        from: '/some/other/path/suffix2',
        to: '/',
      },

      {
        from: '/testPath/some/path/suffix1',
        to: '/testPath',
      },
      {
        from: '/testPath/some/other/path/suffix2',
        to: '/testPath',
      },

      {
        from: '/otherPath.html/some/path/suffix1',
        to: '/otherPath.html',
      },
      {
        from: '/otherPath.html/some/other/path/suffix2',
        to: '/otherPath.html',
      },
    ]);
  });

  it('allows returning string / undefined', () => {
    expect(
      collectRedirects(
        createTestPluginContext(
          {
            createRedirects: (routePath) => {
              if (routePath === '/') {
                return `${routePath}foo`;
              }
              return undefined;
            },
          },
          ['/', '/testPath', '/otherPath.html'],
        ),
        undefined,
      ),
    ).toEqual([{from: '/foo', to: '/'}]);
  });

  it('throws if redirect creator creates invalid redirects', () => {
    expect(() =>
      collectRedirects(
        createTestPluginContext(
          {
            createRedirects: (routePath) => {
              if (routePath === '/') {
                return [
                  `https://google.com/`,
                  `//abc`,
                  `/def?queryString=toto`,
                ];
              }
              return undefined;
            },
          },
          ['/'],
        ),
        undefined,
      ),
    ).toThrowErrorMatchingSnapshot();
  });

  it('throws if redirect creator creates array of array redirect', () => {
    expect(() =>
      collectRedirects(
        createTestPluginContext(
          {
            // @ts-expect-error: for test
            createRedirects(routePath) {
              if (routePath === '/') {
                return [[`/fromPath`]];
              }
              return undefined;
            },
          },
          ['/'],
        ),
        undefined,
      ),
    ).toThrowErrorMatchingSnapshot();
  });

  it('filters unwanted redirects', () => {
    expect(
      collectRedirects(
        createTestPluginContext(
          {
            fromExtensions: ['html', 'exe'],
            toExtensions: ['html', 'exe'],
          },
          [
            '/',
            '/somePath',
            '/somePath.html',
            '/somePath.exe',
            '/fromShouldWork.html',
            '/toShouldWork',
          ],
        ),
        undefined,
      ),
    ).toEqual([
      {
        from: '/toShouldWork.html',
        to: '/toShouldWork',
      },
      {
        from: '/toShouldWork.exe',
        to: '/toShouldWork',
      },
      {
        from: '/fromShouldWork',
        to: '/fromShouldWork.html',
      },
    ]);
  });

  it('throws when creating duplicate redirect routes and onDuplicateRoutes=throw', () => {
    expect(() =>
      collectRedirects(
        createTestPluginContext(
          {
            createRedirects() {
              return '/random-path';
            },
          },
          ['/path-one', '/path-two'],
          {onDuplicateRoutes: 'throw'},
        ),
        undefined,
      ),
    ).toThrowErrorMatchingInlineSnapshot(`
      "@docusaurus/plugin-client-redirects: multiple redirects are created with the same "from" pathname: "/random-path"
      It is not possible to redirect the same pathname to multiple destinations:
      - {"from":"/random-path","to":"/path-one"}
      - {"from":"/random-path","to":"/path-two"}"
    `);
    expect(() =>
      collectRedirects(
        createTestPluginContext(
          {
            redirects: [
              {from: '/path-three', to: '/path-one'},
              {from: '/path-two', to: '/path-one'},
            ],
          },
          ['/path-one', '/path-two'],
          {onDuplicateRoutes: 'throw'},
        ),
        undefined,
      ),
    ).toThrowErrorMatchingInlineSnapshot(`
      "@docusaurus/plugin-client-redirects: some redirects would override existing paths, and will be ignored:
      - {"from":"/path-two","to":"/path-one"}"
    `);
  });
});
