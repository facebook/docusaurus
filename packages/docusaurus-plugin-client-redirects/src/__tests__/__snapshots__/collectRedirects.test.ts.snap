// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collectRedirects throw if plugin option redirects contain invalid to paths 1`] = `
"You are trying to create client-side redirections to invalid paths.

These paths are redirected to but do not exist:
- /this/path/does/not/exist2
- /this/path/does/not/exist3
- /this/path/does/not/exist4

Valid paths you can redirect to:
- /
- /someExistingPath
- /anotherExistingPath
"
`;

exports[`collectRedirects throw if plugin option redirects contain to paths with mismatching trailing slash 1`] = `
"You are trying to create client-side redirections to invalid paths.

These paths do exist, but because you have explicitly set trailingSlash=false, you need to write the path without trailing slash:
- /someExistingPath/
"
`;

exports[`collectRedirects throw if plugin option redirects contain to paths with mismatching trailing slash 2`] = `
"You are trying to create client-side redirections to invalid paths.

These paths do exist, but because you have explicitly set trailingSlash=true, you need to write the path with trailing slash:
- /someExistingPath
"
`;

exports[`collectRedirects throws if redirect creator creates array of array redirect 1`] = `
"Some created redirects are invalid:
- {"from":["/fromPath"],"to":"/"} => Validation error: "from" must be a string
"
`;

exports[`collectRedirects throws if redirect creator creates invalid redirects 1`] = `
"Some created redirects are invalid:
- {"from":"https://google.com/","to":"/"} => Validation error: "from" (https://google.com/) is not a valid pathname. Pathname should start with slash and not contain any domain or query string.
- {"from":"//abc","to":"/"} => Validation error: "from" (//abc) is not a valid pathname. Pathname should start with slash and not contain any domain or query string.
- {"from":"/def?queryString=toto","to":"/"} => Validation error: "from" (/def?queryString=toto) is not a valid pathname. Pathname should start with slash and not contain any domain or query string.
"
`;
